apex/contrib/sparsity/permutation_search_kernels/permutation_utilities.py

```diff
@@ -0,0 +1,137 @@
+import numpy as np
+import time
+import ctypes
+import subprocess
+import os
+import math
+
+gpus_tested = False
+gpus_found = 0
+E = None
+def use_gpu(initial_override = True): 
+    global gpus_tested, gpus_found, E
+    if not gpus_tested:
+        if not initial_override:
+            gpus_tested = True
+            return False, None
+
+        try:
+            gpus_found = str(subprocess.check_output(["nvidia-smi", "-L"])).count('UUID')
+        except:
+            gpus_found = 0
+
+        gpus_tested = True
+        
+        if gpus_found > 0:
+            # Method No.1
+            #script_dir = os.path.dirname(__file__)
+            ##print(script_dir)    # e.g. /opt/conda/lib/python3.8/site-packages/apex/contrib/sparsity/permutation_search_kernels
+            #module_path = os.path.abspath(os.path.join(script_dir, '..', '..', '..', '..'))
+            ##print(module_path)    # e.g. /opt/conda/lib/python3.8/site-packages
+            #python_version_string = os.path.split(os.path.abspath(os.path.join(module_path, '..')))[1]
+            ##print(python_version_string)    # e.g. python3.8
+            #python_version_major_num = python_version_string.split('.')[0][-1]
+            ##print(python_version_major_num)    # e.g. 3
+            #python_version_minor_num = python_version_string.split('.')[1]
+            ##print(python_version_minor_num)    # e.g. 8
+            #cuda_kernel_name = 'permutation_search_cuda.cpython-{:}{:}-x86_64-linux-gnu.so'.format(python_version_major_num, python_version_minor_num)
+            ##print(cuda_kernel_name)    # e.g. permutation_search_cuda.cpython-38-x86_64-linux-gnu.so
+            #cuda_kernel_path = os.path.join(module_path, cuda_kernel_name)    # e.g. /opt/conda/lib/python3.8/site-packages/permutation_search_cuda.cpython-38-x86_64-linux-gnu.so
+
+            # Method No.2, More robust
+            import sysconfig
+            lib_dir = sysconfig.get_config_var("LIBDEST")    # e.g. '/opt/conda/lib/python3.8'
+            #lib_dir = sysconfig.get_config_var("BINLIBDEST")    # e.g. '/opt/conda/lib/python3.8'
+            #print(lib_dir)
+            ext_suffix = sysconfig.get_config_var('EXT_SUFFIX')    # e.g. '.cpython-38-x86_64-linux-gnu.so'
+            #print(ext_suffix)
+            cuda_kernel_name = 'permutation_search_cuda' + ext_suffix
+            #print(cuda_kernel_name)    # e.g. permutation_search_cuda.cpython-38-x86_64-linux-gnu.so
+            cuda_kernel_path = os.path.join(lib_dir, 'site-packages', cuda_kernel_name)    # e.g. /opt/conda/lib/python3.8/site-packages/permutation_search_cuda.cpython-38-x86_64-linux-gnu.so
```

Comment:
This is an invalid path, as e.g. the installation in a conda base environment would point to:
```
/opt/conda/lib/python3.8/site-packages/apex-0.1-py3.8-linux-x86_64.egg/permutation_search_cuda.cpython-38-x86_64-linux-gnu.so
```